### PR TITLE
Msg test/fixes/improvements. Perf test improvements.

### PR DESCRIPTION
**Pull request by drewnoakes**
Wrote some unit tests for the `Msg` class. Uncovered a few glitches, most notably that `Copy` didn't correctly track the reference count for the data buffer. I believe this is fixed now, and if not then someone can document the expected behaviour in a unit test.

The `PerformanceTests/NetMQ.SimpleTests` project got a bit of an overhaul. Its output is easier to read now, and it runs versions of its latency and throughput tests using the `Msg` class. Results of a release build on my Intel i7-2670 2.2GHz laptop:

```text
========  Req/Rep Latency Benchmark

 Iterations: 20,000

 Size   Latency (µs)
---------------------
 8         59.2
 64        56.4
 512       55.7
 4096      70.7
 8192      97.4
 16384    123.0
 32768    163.1

========  Req/Rep Latency Benchmark (reusing Msg)

 Iterations: 20,000

 Size   Latency (µs)
---------------------
 8         56.0
 64        56.2
 512       55.6
 4096      70.2
 8192      93.7
 16384    104.9
 32768    152.2

========  Push/Pull Throughput Benchmark

 Messages: 1,000,000

 Size     Msgs/sec     Mb/s
----------------------------
 8       1144399.3    73.24
 64       866801.7   443.80
 256      438572.5   898.20
 1024     174541.7  1429.85
 4096      44242.4  1449.74

========  Push/Pull Throughput Benchmark (reusing Msg)

 Messages: 1,000,000

 Size     Msgs/sec     Mb/s
----------------------------
 8       1346662.9    86.19
 64      1431994.0   733.18
 256     1401216.1  2869.69
 1024    1417103.6 11608.91
 4096    1443316.5 47294.59
```

These all use the TCP transport.

### CODE REVIEW

**Comment by JamesWHurst** [comment]
Uhm..  about that Push/Pull benchmark...   ?

**Comment by drewnoakes** [comment]
@JamesWHurst, re: benchmark. Looks like front page news for the project. Would be good to spend some time to verify some of these numbers before publishing them. What output does the code give on your machine?

**Comment by lufka** [comment]
Here are some results from another computer (I'm a workmate of @drewnoakes):

```text
======== Req/Rep Latency Benchmark

 Iterations: 20,000

 Size   Latency (µs)
---------------------
 8         59.6
 64        60.3
 512       60.1
 4096      80.1
 8192     128.2
 16384    164.2
 32768    233.1

======== Req/Rep Latency Benchmark (reusing Msg)

 Iterations: 20,000

 Size   Latency (µs)
---------------------
 8         57.6
 64        57.5
 512       60.2
 4096      79.8
 8192     121.6
 16384    156.7
 32768    222.7

======== Push/Pull Throughput Benchmark

 Messages: 1,000,000

 Size     Msgs/sec     Mb/s
----------------------------
 8       1008912.6    64.57
 64       702387.2   359.62
 256      382470.6   783.30
 1024     102951.0   843.37
 4096      25780.1   844.76

======== Push/Pull Throughput Benchmark (reusing Msg)

 Messages: 1,000,000

 Size     Msgs/sec     Mb/s
----------------------------
 8        757017.8    48.45
 64      1166414.3   597.20
 256     1314196.3  2691.47
 1024    1236031.2 10125.57
 4096    1132075.0 37095.83
```

Laptop specs: Intel x64 i5-3320M 2.6GHz

File: src/NetMQ/zmq/SocketBase.cs [hunk 1]
```diff
@@ -768,17 +758,17 @@ public void Send(ref Msg msg, SendReceiveOptions flags)
...
-                        throw new AgainException(innerException: null, message: "SocketBase.Send failed and timeout <= 0");
-                    }
-                }
+                if (timeout <= 0)
+                    continue;
+
+                timeout = (int)(end - Clock.NowMs());
+                    
+                if (timeout <= 0)
+                    throw new AgainException(innerException: null, message: "SocketBase.Send failed and timeout <= 0");
```
Comment (somdoron) [inline src/NetMQ/zmq/SocketBase.cs]: once we will change the API to Try semantic we can also return false instead of throwing exception.

**Comment by somdoron** [comment]
great job. It always nice to see the performance numbers of NetMQ.

We should also think how we can improve the API to reuse the Msg class, currently the API create a new msg on every method call.

We can try and push ZProto which use the Msg to serialize the data and doesn't create a lot of memory.

**Pull request merged**